### ISSUE TITLE
Refine the codebase

### ISSUE THREAD

**Issue by johnslavik** [coherent-oss/python-kustomize#6]

**Comment by jaraco** [comment]
I've been doing this lately with `ruff --select UP --fix`. See also https://github.com/jaraco/skeleton/issues/112, where we're looking to apply this constraint across all projects.

**Comment by diogobaeder** [comment]
I'm all for automating the process, as long as we can have human reviews on the changes (e.g. PRs or review-before-push or so)

**Comment by bswck** [comment]
I don't see much to do here when it comes to Python 3.8 modernization precisely.
I'll then go ahead and say I'll generically improve the code quality.

**Comment by bswck** [comment]
I'll start from type hints.

**Comment by bswck** [comment]
> For example, in 3.6 dataclasses was an external library, and since 3.8 (IIRC) they became part of the stdlib. So we could cut out any reference to dataclasses as a library.

Yeah, dataclasses was added to stdlib in 3.7. But I can see it between other stdlib imports here:
https://github.com/coherent-oss/python-kustomize/blob/a0c5b3c22a5cecd2704a47df87e0aeb00a094607/kustomize/generators.py#L6-L8

And I don't see a place where dataclasses are referred to as if it was a third-party package.

**Comment by diogobaeder** [comment]
Here's a few ideas to look for:

- https://docs.python.org/3/whatsnew/3.7.html#whatsnew37-pep563 - import future annotations and use where classes reference themselves
- https://docs.python.org/3/whatsnew/3.8.html#assignment-expressions - "walrus operator", where it makes sense
- https://docs.python.org/3/whatsnew/3.8.html#positional-only-parameters - positional-only parameters in method headers that could be confusing without them (if this change is implemented, it makes the library's API backwards-incompatible, so we need to make sure to release at least a minor version next time if this happens)
- https://docs.python.org/3/whatsnew/3.8.html#f-strings-support-for-self-documenting-expressions-and-debugging - use f-strings where it makes sense

Other than this, I would be happy if we can make sure the package is mypy-strict. For some years I've been trying hard to keep type hints as strict as possible, because they're immensely helpful.

**Comment by bswck** [comment]
Thanks for assistance. I've been doing such refactors from time to time (https://github.com/CERT-Polska/malduck/pull/111, https://github.com/python-poetry/cleo/pull/365), I'm sure we'll get as much as we can out of it. :)

**Comment by bswck** [comment]
> [docs.python.org/3/whatsnew/3.7.html#whatsnew37-pep563](https://docs.python.org/3/whatsnew/3.7.html#whatsnew37-pep563) - import future annotations and use where classes reference themselves

I'm very happy you don't oppose to PEP 563. Opinions tend to be pretty controversial.
